backend/src/apiserver/storage/pipeline_store_kubernetes.go

```diff
@@ -0,0 +1,552 @@
+	k8sPipeline := v2beta1.Pipeline{}
+
+	err := k.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, &k8sPipeline)
+	if k8serrors.IsNotFound(err) {
+		return nil, util.NewResourceNotFoundError("Namespace/Pipeline", fmt.Sprintf("%v/%v", namespace, name))
+	} else if err != nil {
+		return nil, util.NewInternalServerError(
+			err, "Failed to get a pipeline with name %v and namespace %v", name, namespace,
+		)
+	}
+
+	return k8sPipeline.ToModel(), nil
+}
+
+func (k *PipelineStoreKubernetes) ListPipelinesV1(filterContext *model.FilterContext, opts *list.Options) ([]*model.Pipeline, []*model.PipelineVersion, int, string, error) {
+	return nil, nil, 0, "", ErrNoV1
+}
+
+func (k *PipelineStoreKubernetes) ListPipelines(filterContext *model.FilterContext, opts *list.Options) ([]*model.Pipeline, int, string, error) {
+	k8sPipelines := v2beta1.PipelineList{}
+
+	listOptions := []ctrlclient.ListOption{ctrlclient.UnsafeDisableDeepCopy}
+
+	if filterContext.ReferenceKey != nil && filterContext.ReferenceKey.Type == model.NamespaceResourceType {
+		listOptions = append(listOptions, ctrlclient.InNamespace(filterContext.ReferenceKey.ID))
+	}
+
+	// Be careful, the deep copy is disabled here to reduce memory allocations
+	err := k.client.List(context.TODO(), &k8sPipelines, listOptions...)
+	if err != nil {
+		return nil, 0, "", util.NewInternalServerError(
+			err, "Failed to find the pipeline associated with this pipeline version",
+		)
+	}
+
+	pipelines := make([]*model.Pipeline, 0, len(k8sPipelines.Items))
+
+	for _, k8sPipeline := range k8sPipelines.Items {
+		if opts.Filter == nil {
+			pipelines = append(pipelines, k8sPipeline.ToModel())
+			continue
+		} else {
+			found, err1 := opts.Filter.FilterK8sPipelines(k8sPipeline)
+			if err1 != nil {
+				return nil, 0, "", err1
+			}
+			if found {
+				pipelines = append(pipelines, k8sPipeline.ToModel())
+			}
+		}
+	}
+
+	// Because controller-client does not have sorting, use this function to sort by fields.
+	sort.Slice(pipelines, func(i, j int) bool {
+		if opts.SortByFieldName != "" {
+			elementA := pipelines[i].GetFieldValue(opts.SortByFieldName)
+			elementB := pipelines[j].GetFieldValue(opts.SortByFieldName)
+
+			if elementA != nil && elementB != nil {
+				switch elementA.(type) {
+				case int64:
+					return elementA.(int64) > elementB.(int64)
+				case float64:
+					return elementA.(float64) > elementB.(float64)
+				case string:
+					return elementA.(string) > elementB.(string)
+				default:
+					glog.Warningf("Field type %T in %s not recognized. Sorting will not work.", elementA, opts.SortByFieldName)
+					return false
+				}
+			}
+		}
+		return false
+	})
+	if !opts.IsDesc {
+		slices.Reverse(pipelines)
+	}
+
+	// If there's pagination, find the index of the first element of the next page
+	nextPage := 0
+	for p, pipeline := range pipelines {
+		// If opts.KeyFieldValue is empty, skip the loop
+		if opts.SortByFieldValue == "" {
```

Comment:
```suggestion
		if opts.KeyFieldValue == "" {
```

Reply:
In general, I prefer the approach taken in ListPipelineVersions where the for loop is inside the if.